Add support for WebGL2 TexImage2D

imiklos opened the pull request:
Adds initial support for one of the WebGL2 `TexImage2D` call.

I've enabled the `tests/wpt/webgl/tests/deqp/` tests.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix part of #26512 
- [x] There are tests for these changes 

@mmatyas @zakorgy @jdm 
<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Commit:
Add support for WebGL2 TexImage2D

Adds initial support for one of the WebGL2 `TexImage2D` call.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webgl2renderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl
 * @kichjang: components/script/dom/webgl2renderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl

imiklos commented:
I'm getting `INVALID_OPERATION` from the `GL` side on every call. I'm stuck and not sure about the current implementation.
Looking for some advice.

jdm commented:
> I'm getting `INVALID_OPERATION` from the `GL` side on every call. I'm stuck and not sure about the current implementation.
> Looking for some advice.

The reason this occurs is the pixel unpack buffer is bound when we invoke glTexImage2d. This causes us to treat the pointer passed as the pixel data buffer to be used as an offset into the bound pixel unpack buffer, instead.

The following patch works around this:
```diff
diff --git a/components/canvas/webgl_thread.rs b/components/canvas/webgl_thread.rs
index 9b7b4ed468..da9a502d1d 100644
--- a/components/canvas/webgl_thread.rs
+++ b/components/canvas/webgl_thread.rs
@@ -1586,6 +1586,11 @@ impl WebGLImpl {
                 );

                 gl.pixel_store_i(gl::UNPACK_ALIGNMENT, unpacking_alignment as i32);
+                let mut buffer = [0];
+                unsafe {
+                    gl.get_integer_v(gl::PIXEL_UNPACK_BUFFER_BINDING, &mut buffer);
+                }
+                gl.bind_buffer(gl::PIXEL_UNPACK_BUFFER, 0);
                 gl.tex_image_2d(
                     target,
                     level as i32,
@@ -1597,6 +1602,7 @@ impl WebGLImpl {
                     effective_data_type,
                     Some(&pixels),
                 );
+                gl.bind_buffer(gl::PIXEL_UNPACK_BUFFER, buffer[0] as u32);
             },
             WebGLCommand::TexSubImage2D {
                 target,
```

However, it would be more efficient to use a different model - instead of retrieving the pixels from the unpack buffer, then sending those pixels to be treated like any other 2d texture, I think we should make the WebGLRenderingContext::tex_image_2d take an enum argument like:
```rust
enum TexSource {
    Pixels(TexPixels),
    BufferOffset(i64),
}
```
and modify WebGLThread to handle that case appropriately.

imiklos commented:
@jdm Thank you! I've updated the pull request with the changes you mentioned, also updated the test expectations according to the ci. This pull request is depends on https://github.com/servo/sparkle/pull/28.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 9949c5d has been approved by `jdm`

<!-- @bors-servo r=jdm 9949c5d8ca149b0873802e26b4749bfb2211fd06 -->

bors-servo commented:
:hourglass: Testing commit 9949c5d8ca149b0873802e26b4749bfb2211fd06 with merge 0c059dd187cdef5d7945145f032344325113a427...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/A9V8sgtMSE6tVyHX1ojmAA)

jdm commented:
Cargo.lock needs to be updated with the latest package information:
```
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -5311,9 +5311,9 @@ dependencies = [
 
 [[package]]
 name = "sparkle"
-version = "0.1.24"
+version = "0.1.25"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b7cd7f292661a533c2a7e239af495ba63891b5723c3a6d48712b2e3823466174"
+checksum = "7184285ed8ddbf106cda8fef17f584207fc64a03125e3694c324b19b708825fa"
 dependencies = [
  "gl_generator 0.13.1",
 ]
```

imiklos commented:
I've updated the `Cargo.lock`.

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit 55a93d4 has been approved by `jdm`

<!-- @bors-servo r=jdm 55a93d447b491ae5da6a908b7c2981e36e98e2cc -->

bors-servo commented:
:hourglass: Testing commit 55a93d447b491ae5da6a908b7c2981e36e98e2cc with merge a53e56b782bd9b336091ab9043bdd9c406c31f9e...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/HFhGjWwpQoSZKr_fMAp3vg)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26697) made this pull request unmergeable. Please resolve the merge conflicts.

imiklos commented:
I've rebased the pull request to the latest master and updated the test expectations.

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit a5916db has been approved by `jdm`

<!-- @bors-servo r=jdm a5916dbe9c99c5abba9b3a6032c23d39d4fc092c -->

bors-servo commented:
:hourglass: Testing commit a5916dbe9c99c5abba9b3a6032c23d39d4fc092c with merge bc0f77c2f9e110bcb8201a8d663adeb89c7afd83...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/KnX-t1DPQim409xrPAg3jA)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26872) made this pull request unmergeable. Please resolve the merge conflicts.

imiklos commented:
I've rebased the pull request and updated the test expectations.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit b477129c1ae920f49632b8fbc0b35e79d9f9035c with merge ae18a22175d452ff58ab3f9682dbc1617ee0ebeb...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/S3zJMntcRVeLD16YBHLqtA)

jdm commented:
I fear the deqp tests may be too inconsistent to enable :/

imiklos commented:
I've disabled the `deqp` tests.

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit 6591fa5 has been approved by `jdm`

<!-- @bors-servo r=jdm 6591fa54f929dd76a0383f5b40a7490f22d0f7cc -->

bors-servo commented:
:hourglass: Testing commit 6591fa54f929dd76a0383f5b40a7490f22d0f7cc with merge 6b5455f90195065097337c430580fc222e78b0cc...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/B-ATbhvzQ-an_em_ckDJLQ)

jdm commented:
@bors-servo retry
* #24622

bors-servo commented:
:hourglass: Testing commit 6591fa54f929dd76a0383f5b40a7490f22d0f7cc with merge 1b55ab5804d99f54fd45b3a9ec4a20154b38bf9b...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/aqK9tyDIQcKVjRpS37VMQA)
Approved by: jdm
Pushing 1b55ab5804d99f54fd45b3a9ec4a20154b38bf9b to master...
